docs/proposals/VIP-2.md

```diff
@@ -0,0 +1,240 @@
+#### **Leap Frog**
+
+Leap Frog works on the notion that we can detect if two replicas diverge for a record if the following holds true when comparing replica A and B
+
+1. The value for a certain key on replicas A and B are different AND
+2. The record on replica A was received from an offset that falls below a highwatermark of applied events on replica B AND
+3. The value on replica B was received from an offset that falls below a highwatermark of applied events on replica A
+
+If all three are true, then we can detect divergence.  The implementation of LeapFrog is based on the consumption of a Venice version topic (as opposed to interrogating the on disk state of Venice).
+
+We won't deep dive Leap Frog here, but in order to give the reader more context, there is [a TLA+ spec included with this PR](../../specs/TLA%2B/LeapFrog/leapfrog.tla) which explains and and provides some validation of the correctness of the detection algorithm.
+
+#### **Venice change capture**
+
+Venice change capture relies on per record metadata in order for the client to be able to screen out duplicate events.  Duplicated events either come in the form of duplicate events sent from the PubSub Broker, or events which are applied following a rewind on a version push.  It works by maintaining a high watermark state on the client which screens out events which fall below the cached high watermark.
+
+#### How are offsets stored
+
+When a record is consumed from a given colo's RT, the offset position that it was consumed is noted and merged into a vector of offsets stored alongside that record.  The vector is a set of high watermark offsets consumed from each replicating site. So if we have three sites actively replicating a given partition, we'll store at most three offsets per record. We store this because of the way Venice handles write compute and AA.  The state of the record can be the accumulated state of disjoint updates from different colos, and this is why it's not sufficient to store just the LAST update to a given record.
+
+### What are the base common requirements
+
+If you squint at both of these, they boil down to relying on the following properties.
+
+* Being able to build a high watermark based on events consumed
+* Being able to causally compare a consumed event/record to this highwatermark and take action.
+
+We're able to meet the first requirement by just starting at an arbitrary point in a version topic stream and consuming.  Every single event has an RMD field that can help a consumer of this data build a highwatermark.  And then since every single record we consume has the same information, it's very straightforward to determine the causality of that event relative to a built up high watermark.
+
+That said, it's not actually a requirement to be able to do this on every single event we consume. It's possible to meet the first two requirements at a courser granularity of updates.
+
+### Heartbeat Algorithm
```

Comment:
Not sure my understanding is correct or not, just want to confirm: If we have 3 regions, the Heartbeat algorithm can save RMD space from storing 3 regions's offset to 1 region's offset?

Reply:
You are correct, but we'll actually store none of the regions offset metadata in rocksdb.  We'll only persist it to PubSubBroker as envelope metadata.  None of which has to go into rocksdb.